modules/acton/src/restate/services/LeadVirtualObject.ts

```diff
@@ -0,0 +1,176 @@
+import type { UUID } from "node:crypto";
+import * as restate from "@restatedev/restate-sdk";
+import type { ExternalIntegrationState } from "common/external";
+import { serializeError } from "serialize-error";
+import { assert, is } from "tsafe";
+import { z } from "zod";
+import { SyncWithDB } from "../../db";
+import { WebLeadIntegrations } from "../../external/index";
+import {
+	ActOnLeadSchema,
+	type LeadState,
+	type WebFormLead,
+	type WebLead,
+} from "../../types";
+
+/**
+ * Helper function that runs before all of our exclusive handlers
+ * Handles initializing state from the database verifying assumptions
+ * @param ctx the restate Object context
+ * @param allowedStates a set of lead status states that we should allow this handler to run with
+ */
+async function setup(
+	ctx: restate.ObjectContext<LeadState>,
+	allowedStates: LeadState["Status"][],
+) {
+	const uuidParser = z.string().uuid();
+	if (!uuidParser.safeParse(ctx.key).success) {
+		throw new restate.TerminalError("Lead ID is not a valid UUIDv4", {
+			errorCode: 400,
+		});
+	}
+	if ((await ctx.get("Status")) === "ERROR") {
+		ctx.clearAll();
+	}
+	try {
+		await SyncWithDB(ctx, "RECEIVE");
+	} catch (e) {
+		await ctx.update((_) => ({
+			Status: "ERROR",
+			Error: serializeError(e),
+		}));
+		throw e;
+	}
+
+	const status = (await ctx.get("Status")) ?? "NONEXISTANT";
+	const canRun = allowedStates.includes(status);
+	if (!canRun) {
+		throw new restate.TerminalError(
+			`Lead ID '${ctx.key}' doesn't exist or is in incorrect status '${status}'`,
+			{ errorCode: 409 },
+		);
+	}
+}
+
+export const LeadVirtualObject = restate.object({
+	name: "Lead",
+	handlers: {
+		status: restate.handlers.object.shared(
+			async (
+				ctx: restate.ObjectSharedContext<LeadState>,
+				req: Record<string, any>,
+			): Promise<LeadState> => {
+				//await CheckAuthorization(ctx);
+				const state = await ctx.getAll();
+				if (state.Status == null) {
+					return { Status: "NONEXISTANT" };
+				}
+				return state;
+			},
+		),
+		/**
+		 * Creates a new lead in the database
+		 */
+		create: restate.handlers.object.exclusive(
+			async (
+				ctx: restate.ObjectContext<LeadState>,
+				req: Record<string, any>,
+			): Promise<LeadState> => {
+				try {
+					await ctx.update((_) => ({
+						Status: "VALIDATING",
+						Request: req ?? {},
+					}));
+					// Validate the submitted lead
+					const Lead = ActOnLeadSchema.parse(req);
```

Comment:
Will want to format this so that end users see a human-friendly error
```suggestion
					const parseRequest = ActonLeadSchema.safeParse(req);
					if (!parseRequest.success) {
						const formattedError = fromError(parseRequest.error);
						throw new restate.TerminalError(
							`Request could not be parsed - ${formattedError.message}`,
							{
								errorCode: 400,
							},
						);
					}
```